### PR TITLE
Track source locations on Daml.Script.submit

### LINKED ISSUE #7192: Source locations of commits in DAML Script
At least in DAML Script in DAML Studio, ideally also on failures in the command line runner, we should have the location of the `submit` like we do in scenarios.

I believe in scenarios this is accomplished by matching on `ELocation(loc, EScenario(scen))` in the compiler. This obviously doesn’t work but I think the location might be somewhere in the stack trace.

### PR DESCRIPTION
**Pull request by aherrmann-da**
Closes #7192

Tracks source locations on `Daml.Script.submit` and `Daml.Script.submitMustFail` such that the transaction view in the IDE contains clickable links to the source location of the commit.

Adds a `HasCallStack` constraint on `HasSubmit.submit` and `HasSubmit.submitMustfail`.

We need to move the definition of `HasCallStack` to `GHC.Stack.Types` to avoid a cyclic dependency between `DA.Stack` and `DA.Internal.LF`. Alternatively, we could inline `HasCallStack` in `DA.Internal.LF` but it seems preferable to use the alias on the signature of `submit(MustFail)`.

We need to adapt the type of the builtins for scenario `submit` and `submitMustfail`.

Adds test cases to the lsp-tests that check for the presence of source locations. Adds tests for both DAML Script and scenarios for completeness.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala [hunk 1]
```diff
@@ -383,6 +383,58 @@ object Converter {
...
+                  for {
+                    packageId <- toText(vals.get(0)).flatMap {
+                      // GHC uses unit-id "main" for the current package,
+                      // but the scenario context expects "-homePackageId-".
+                      case "main" => PackageId.fromString("-homePackageId-")
+                      case id => knownPackages.get(id).toRight(s"Unknown package $id")
+                    }
+                    module <- toText(vals.get(1)).flatMap(ModuleName.fromString(_))
+                    // Lines and columns are 1-based in the call stack.
+                    startLine <- toInt(vals.get(3))
```
Comment (cocreature) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:419]: hm, I wonder if we should make them zero-based in our custom `SrcLoc`. Probably better to follow the LSP convention than GHC and do the conversion at the beginning.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala [hunk 2]
```diff
@@ -380,7 +386,12 @@ class Runner(compiledPackages: CompiledPackages, script: Script.Action, timeMode
                       client <- Converter.toFuture(
                         clients
                           .getPartyParticipant(Party(party.value)))
-                      submitRes <- client.submit(party, commands)
+                      commitLocation <- if (vals.size == 3) {
```
Comment (cocreature) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala:389]: maybe worth adding a comment that this is for backwards compat.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala [hunk 3]
```diff
@@ -441,7 +452,12 @@ class Runner(compiledPackages: CompiledPackages, script: Script.Action, timeMode
                       client <- Converter.toFuture(
                         clients
                           .getPartyParticipant(Party(party.value)))
-                      submitRes <- client.submitMustFail(party, commands)
+                      commitLocation <- if (vals.size == 3) {
```
Comment (cocreature) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala:455]: same here.

**Review by cocreature — APPROVED** [review]
Fantastic, thank you very much!

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala (hunk 2 above)
Comment (aherrmann-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala:389]: done

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala (hunk 3 above)
Comment (aherrmann-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala:455]: done

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala (hunk 1 above)
Comment (aherrmann-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:420]: Fair enough, I've made the user facing `SrcLoc` zero based.

File: compiler/lsp-tests/src/Main.hs [hunk 4]
```diff
@@ -527,6 +539,21 @@ scriptTests damlcPath scriptDarPath = testGroup "scripts"
...
+    , testCase "submit location" $ run $ do
+          main' <- openDoc' "Main.daml" damlId $ T.unlines
+              [ "{-# LANGUAGE ApplicativeDo #-}"
+              , "module Main where"
+              , "import Daml.Script"
+              , "template T with party : Party where signatory party"
+              , "main : Script (ContractId T)"
+              , "main = do"
+              , "  alice <- allocateParty \"Alice\""
+              , "  submit alice do createCmd (T alice)"
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/lsp-tests/src/Main.hs:553]: Agreed, the script behavior is better :+1:

**Pull request merged**